storage_controller/src/service/safekeeper_reconciler.rs

```diff
@@ -230,32 +245,42 @@ impl SafekeeperReconciler {
     }
     async fn run(&mut self) {
         loop {
-            // TODO add parallelism with semaphore here
             let req = tokio::select! {
                 req = self.rx.recv() => req,
                 _ = self.cancel.cancelled() => break,
             };
             let Some((req, req_cancel)) = req else { break };
+
+            let inner = self.inner.clone();
             if req_cancel.is_cancelled() {
                 continue;
             }
 
-            let kind = req.kind;
-            let tenant_id = req.tenant_id;
-            let timeline_id = req.timeline_id;
-            let node_id = req.safekeeper.skp.id;
-            self.reconcile_one(req, req_cancel)
-                .instrument(tracing::info_span!(
-                    "reconcile_one",
-                    ?kind,
-                    %tenant_id,
-                    ?timeline_id,
-                    %node_id,
-                ))
-                .await;
+            tokio::task::spawn(async move {
+                let kind = req.kind;
+                let tenant_id = req.tenant_id;
+                let timeline_id = req.timeline_id;
+                let node_id = req.safekeeper.skp.id;
```

Comment:
Is there any logic that prevents two on-going reconciliations of the same timeline?

Reply:
`ReconcilerHandle::new_token_slot` cancels previous reconciliations.

But looking at the code, we seem to not clean up `ongoing_tokens`...

Reply:
We fire `.cancel()` on the previous reconciliation token there, but I'm unsure if it's sufficient. If reconcile does some work between checking the cancelation token (I didn't read this code yet), these steps may be run in parallel with the new reconciliation.
Could you please check if it is the problem in our case?

Reply:
The per-task cancellation token is is passed to `reconcile_inner` which gives it to `with_client_retries`, which passes it to `backoff::retry`. That function then checks it each time it does a retry. Ongoing requests are not interrupted, but retries are.

I'd handle this one layer below at the safekeeper level: in general, one needs to design cloud systems in a way that "outdated" requests don't cause problems. There is many ways for such outdated requests to arrive at safekeepers. That's why there is generations, because they give safekeepers a piece of information on when to reject outdated information.